File: apps/share-backend/src/cors.ts
/**
 * CORS wrapper. The frontend is served from a different origin
 * (GitHub Pages) than the Worker (workers.dev or a custom domain),
 * so we have to opt the browser in explicitly. Allowed origins are
 * declared in `wrangler.jsonc#vars.ALLOWED_ORIGINS` so the surface
 * is locked down per deploy rather than baked into code.
 */

const CORS_METHODS = "GET, POST, OPTIONS";
const CORS_HEADERS = "content-type";
const CORS_MAX_AGE = "86400";

function parseAllowed(raw: string | undefined): readonly string[] {
	if (!raw) return [];
	return raw
		.split(",")
		.map((s) => s.trim())
		.filter(Boolean);
}

export function originIsAllowed(
	origin: string | null,
	allowedRaw: string | undefined,
): boolean {
	if (!origin) return false;
	return parseAllowed(allowedRaw).includes(origin);
}

export function corsHeadersFor(
	origin: string | null,
	allowedRaw: string | undefined,
): Headers {
	const headers = new Headers();
	if (originIsAllowed(origin, allowedRaw)) {
		headers.set("access-control-allow-origin", origin as string);
		headers.set("vary", "origin");
		headers.set("access-control-allow-methods", CORS_METHODS);
		headers.set("access-control-allow-headers", CORS_HEADERS);
		headers.set("access-control-max-age", CORS_MAX_AGE);
	}
	return headers;
}

export function preflightResponse(
	origin: string | null,
	allowedRaw: string | undefined,
): Response {
	const headers = corsHeadersFor(origin, allowedRaw);
	const status = originIsAllowed(origin, allowedRaw) ? 204 : 403;
	return new Response(null, { status, headers });
}

/** Merge CORS headers onto an existing response without rebuilding it. */
export function withCors(response: Response, cors: Headers): Response {
	if (cors.has("access-control-allow-origin")) {
		const merged = new Headers(response.headers);
		for (const [k, v] of cors) merged.set(k, v);
		return new Response(response.body, {
			status: response.status,
			statusText: response.statusText,
			headers: merged,
		});
	}
	return response;
}

File: apps/share-backend/src/cors.rs
//! CORS allow-list.
//!
//! The frontend (`opfs-web`) and this backend (`opfs-share-backend`)
//! deploy as two separate Knative Services on two different
//! Knative-assigned hostnames, so the browser issues cross-origin
//! POSTs that the allow-list must cover. `ALLOWED_ORIGINS` is the
//! only thing connecting the two — set it per cluster to the
//! frontend's Knative URL.
//!
//! Parsed once at startup from the env, re-checked per request —
//! no runtime mutation.

use axum::extract::State;
use axum::http::{HeaderMap, HeaderValue, Method, Request, StatusCode, header};
use axum::middleware::Next;
use axum::response::Response;

use crate::state::AppState;

const ALLOWED_METHODS: &str = "GET, POST, OPTIONS";
const ALLOWED_HEADERS: &str = "content-type";
const PREFLIGHT_MAX_AGE: &str = "86400";

pub fn origin_is_allowed(origin: &str, allowed: &[String]) -> bool {
    allowed.iter().any(|a| a == origin)
}

/// Tower-style middleware. Runs ahead of every route — including
/// OPTIONS preflight — so the allow-list is honored on simple
/// cross-origin POSTs (which skip preflight).
pub async fn cors_layer(
    State(state): State<AppState>,
    request: Request<axum::body::Body>,
    next: Next,
) -> Response {
    let origin = request
        .headers()
        .get(header::ORIGIN)
        .and_then(|v| v.to_str().ok())
        .map(str::to_owned);

    // Preflight: short-circuit. 204 with CORS headers on allowed
    // origins, 403 otherwise.
    if request.method() == Method::OPTIONS {
        return preflight_response(origin.as_deref(), &state.allowed_origins);
    }

    // Simple cross-origin POSTs bypass preflight, so enforce the
    // allow-list before any handler runs.
    if let Some(ref o) = origin {
        if !origin_is_allowed(o, &state.allowed_origins) {
            return forbidden_origin();
        }
    }

    let mut response = next.run(request).await;
    // Once we reach this point the origin has already passed the
    // pre-handler gate above, so the second `origin_is_allowed`
    // call gemini flagged would be redundant. Just apply the
    // headers when the origin is present.
    if let Some(o) = origin {
        apply_cors_headers(response.headers_mut(), &o);
    }
    response
}

fn preflight_response(origin: Option<&str>, allowed: &[String]) -> Response {
    let ok = origin.is_some_and(|o| origin_is_allowed(o, allowed));
    let status = if ok {
        StatusCode::NO_CONTENT
    } else {
        StatusCode::FORBIDDEN
    };
    let mut response = Response::builder()
        .status(status)
        .body(axum::body::Body::empty())
        .expect("static body");
    if ok {
        // `ok` implies `origin.is_some_and(...)`, so the unwrap is
        // proven by the branch condition.
        apply_cors_headers(response.headers_mut(), origin.expect("ok ⇒ Some"));
    }
    response
}

fn forbidden_origin() -> Response {
    let body = serde_json::json!({ "error": "origin not allowed" });
    Response::builder()
        .status(StatusCode::FORBIDDEN)
        .header(header::CONTENT_TYPE, "application/json")
        .body(axum::body::Body::from(body.to_string()))
        .expect("static body")
}

fn apply_cors_headers(headers: &mut HeaderMap, origin: &str) {
    if let Ok(value) = HeaderValue::from_str(origin) {
        headers.insert(header::ACCESS_CONTROL_ALLOW_ORIGIN, value);
    }
    headers.insert(header::VARY, HeaderValue::from_static("origin"));
    headers.insert(
        header::ACCESS_CONTROL_ALLOW_METHODS,
        HeaderValue::from_static(ALLOWED_METHODS),
    );
    headers.insert(
        header::ACCESS_CONTROL_ALLOW_HEADERS,
        HeaderValue::from_static(ALLOWED_HEADERS),
    );
    headers.insert(
        header::ACCESS_CONTROL_MAX_AGE,
        HeaderValue::from_static(PREFLIGHT_MAX_AGE),
    );
}
